libr/anal/jmptbl.c

```diff
@@ -42,8 +42,105 @@ R_API bool r_anal_jmptbl(RAnal *anal, RAnalFunction *fcn, RAnalBlock *block, ut6
 	return try_walkthrough_jmptbl (anal, fcn, block, depth, jmpaddr, table, table, tablesize, tablesize, default_addr, false);
 }
 
+R_API bool try_walkthrough_casetbl(RAnal *anal, RAnalFunction *fcn, RAnalBlock *block, int depth, ut64 ip, ut64 jmptbl_loc, ut64 casetbl_loc, ut64 jmptbl_off, ut64 sz, ut64 jmptbl_size, ut64 default_case, bool ret0) {
+	bool ret = ret0, default_in_switch = false;
+	if (jmptbl_size == 0) {
+		jmptbl_size = JMPTBL_MAXSZ;
+	}
+	if (jmptbl_loc == UT64_MAX) {
+		if (anal->verbose) {
+			eprintf ("Warning: Invalid JumpTable location 0x%08" PFMT64x "\n", jmptbl_loc);
+		}
+		return false;
+	}
+	if (casetbl_loc == UT64_MAX) {
+		if (anal->verbose) {
+			eprintf ("Warning: Invalid CaseTable location 0x%08" PFMT64x "\n", jmptbl_loc);
+		}
+		return false;
+	}
+	if (jmptbl_size < 1 || jmptbl_size > ST32_MAX) {
+		if (anal->verbose) {
+			eprintf ("Warning: Invalid JumpTable size at 0x%08" PFMT64x "\n", ip);
+		}
+		return false;
+	}
+	ut64 jmpptr, case_idx, jmpptr_idx;
+	ut8 *jmptbl = calloc (jmptbl_size, sz);
+	if (!jmptbl) {
+		return false;
+	}
+	ut8 *casetbl = calloc (jmptbl_size, sizeof (ut8));
+	if (!casetbl) {
+		return false;
+	}
+	anal->iob.read_at (anal->iob.io, jmptbl_loc, jmptbl, jmptbl_size * sz);
+	anal->iob.read_at (anal->iob.io, casetbl_loc, casetbl, jmptbl_size);
+	ut64 case_count = 0;
+	for (case_idx = 0; case_idx < jmptbl_size; case_idx++) {
+		jmpptr_idx = casetbl[case_idx];
+
+		switch (sz) {
+		case 1:
+			jmpptr = r_read_le8 (jmptbl + jmpptr_idx);
+			break;
+		case 2:
+			jmpptr = r_read_le16 (jmptbl + jmpptr_idx * 2);
+			break;
+		case 4:
+			jmpptr = r_read_le32 (jmptbl + jmpptr_idx * 4);
+			break;
+		default:
+			jmpptr = r_read_le64 (jmptbl + jmpptr_idx * 8);
+			break;
+		}
+		if (jmpptr == 0 || jmpptr == UT32_MAX || jmpptr == UT64_MAX) {
```

Comment:
ok I admit i have not read the whole PR yet, but why do you check for jmpptr == 0? Can't it be a valid value?

Reply:
@ret2libc see this comment:
https://github.com/radareorg/radare2/blob/27a56e848689a87e6f3b27855567148419803e88/libr/anal/jmptbl.c#L171-L174